posit-dev/positron
# Jump-to-layout action

Addresses #3040 #2835 and sets up the foundations for #3039 as part of #3038 epic.

### Intent

Add an action that allows the jumping to a pre-specified layout immediately. Starts with three predefined layouts: 4-pane data science, side-by-side data science (with plot), and side-by-side data science (without plot). 

### Approach

A new action was added: `'workbench.action.chooseLayout'`. This action shows a quick-pick menu that allows the user to select from a list of pre-defined layouts. 

The layouts are defined with a JSON structure in the following format. 

```ts
/**
 * Full description of custom layout for the editor.
 */
type CustomPositronLayoutDescription = Record<
	KnownPositronLayoutParts,
	PartLayoutDescription
>;

/**
 * Known parts that can be customized in for layouts.
 */
type KnownPositronLayoutParts = Parts.PANEL_PART | Parts.SIDEBAR_PART | Parts.AUXILIARYBAR_PART;

/**
 * Description of the custom layout for a given part (e.g. Sidebar, Panel, ...) of the editor.
 */
type PartLayoutDescription = {
	/**
	 * Size of the part. If a number, it's an absolute size in pixels. If it's a string it's a
	 * relative size in percentage of the viewport size. If the size controls the width or the
	 * height depends on the part. E.g. for the sidebar it's the width.
	 */
	size?: number | `${number}%`;
	/**
	 * Should this part be hidden by default?
	 */
	hidden: boolean;
	/**
	 * Alignment of the part. Only used for the panel part.
	 */
	alignment?: PanelAlignment;
	/**
	 * Description of the view containers in this part. The order as they appear in the array
	 * will be the order they are shown in the UI. Any non-specified view containers will be
	 * added after the specified ones.
	 */
	viewContainers?: ViewContainerLayoutDescription[];
};

/**
 * Description of a view container within an editor part. E.g. the "Sessions" tab.
 */
type ViewContainerLayoutDescription = {
	/**
	 * Id of this view container. This is the id that the view container is registered with.
	 * E.g. `workbench.panel.positronSession`.
	 */
	id: string;
	/**
	 * Is this view container shown? Only one of these can be shown at a time so if multiple are
	 * set, the last one will be respected.
	 */
	opened?: boolean;
	/**
	 * Description of the views within this view container. The order as they appear in the array
	 * will be the order they are shown in the UI. Any non-specified views will be added after the
	 * specified ones.
	 */
	views?: ViewLayoutDescription[];
};

type ViewLayoutDescription = {
	/**
	 * Id of this view. This is the id that the view is registered with.
	 * E.g. `workbench.panel.positronPlots` or `terminal`.
	 */
	id: string;
	/**
	 * Size units are relative. Every view sharing the same `relativeSize` will have the same size.
	 * If not provided, will default to 1.
	 */
	relativeSize?: number;
};
```

For instance, here's the definition of the 4-pane data science layout: 
```ts
{
	[Parts.SIDEBAR_PART]: {
		size: 200,
		hidden: false,
	},
	[Parts.PANEL_PART]: {
		size: '40%',
		hidden: false,
		alignment: 'center',
		viewContainers: [
			{
				id: 'workbench.panel.positronConsole',
				opened: true,
			},
			{
				id: 'terminal',
			},
		]
	},
	[Parts.AUXILIARYBAR_PART]: {
		size: 500, // Use pixel sizes for auxiliary bar to allow editor to take up the rest of the space
		hidden: false,
		viewContainers: [
			{
				id: 'workbench.panel.positronSession',
                                 opened: true,
				views: [
					{
						id: 'workbench.panel.positronVariables',
					},
					{
						id: 'workbench.panel.positronPlots',
					},
				]
			},
		]
	}
};
```

This layout when run produces the following layout (on my 16inch macbook screen.)
<img width="2056" alt="image" src="https://github.com/posit-dev/positron/assets/6764693/33b20a75-143b-4bb6-97c1-558740820f56">

Other layouts include
__Side-by-side Data Science:__
<img width="2056" alt="image" src="https://github.com/posit-dev/positron/assets/6764693/6454eeea-32ee-4b55-bfb2-53a900ea0177">

and __Side-by-side Data Science (Console-only):__

<img width="2056" alt="image" src="https://github.com/posit-dev/positron/assets/6764693/a15b5c74-17e6-4f87-9ac8-1719f609e7e8">

## Implementation

To enable this two existing services were touched: the `IViewDescriptorService` where we added a `loadCustomViewDescriptor()` method that takes a `CustomPositronLayoutDescription` and rearranges the views and `ILayoutService` where we added `.enterCustomLayout()` and `getPartViewInfo()` which size and hide/show the various panels (e.g. sidebars) and return information on a given panel for use in `enterCustomLayout()` respectively. 

Additionally, we added a few methods to the `ViewPaneContainer` and `PaneCompositeBar` classes so we could get access to resizing and ordering functionality that was previously hidden behind internal drag-and-drop logic or caching. 
 
Also during the development, there was a `dump layout` action that dumped the JSON for the current layout to the console. As the layout structure became more complex it became burdensome to keep this up-to-date so I commented out the logic there. Once we're happy with the format/ feature scope I will re-enable it, so I left the commented-out code in. 


### QA Notes

You should be able to run the action and select a given layout from any state of the editor and Positron should look the same. Views should not disappear or get into weird focus states.
